### PR TITLE
refactor: distinguish between Unique and UniqueIndex

### PR DESCRIPTION
**Pull request by black-06**
- [X] Do only one thing
- [X] Non breaking API changes
- [X] Tested

### What did this pull request do?

See https://github.com/go-gorm/gorm/issues/6381.

based on https://github.com/go-gorm/gorm/pull/6386 and tested

### CODE REVIEW

**Review by a631807682 — APPROVED** [review]
Depends on gorm's PR, this is a breaking change, releases need to pay attention to the version number.

**Comment by black-06** [comment]
It still has some problems which I'm fixing

**Comment by black-06** [comment]
After testing, it works well with https://github.com/go-gorm/gorm/pull/6822

File: migrator.go [hunk 1]
```diff
@@ -47,6 +47,78 @@ func (m Migrator) FullDataTypeOf(field *schema.Field) clause.Expr {
...
+				if uni, ok := index.Unique(); !ok || !uni {
+					continue
+				}
+				if columns := index.Columns(); len(columns) != 1 || columns[0] != field.DBName {
+					continue
+				}
+				if name := index.Name(); name == constraint || name == field.UniqueIndex {
+					continue
+				}
+				if err := execTx.Migrator().DropIndex(value, index.Name()); err != nil {
```
Comment (amodpandey) [inline migrator.go:76]: I had a column for which I changed from `index` to `uniqueIndex`

MyColumn       int32 `gorm:"index"` 
Gorm had created idx_<table name>_<column_name>

which I changed to 

MyColumn       int32 `gorm:"uniqueIndex"`
Gorm retained old index idx_<table name>_<column_name> and created a new unique index `<column_name>`

After this updatae

gorm.io/driver/mysql v1.5.2 => v1.5.6
gorm.io/gorm v1.25.5 => v1.25.10

It dropped index <column_name> which had unique index and tried to create index with name idx_<table name>_<column_name> and failed.

Ideally we should have also dropped idx_<table name>_<column_name> once we realised we are creating it.

As per the text here - https://gorm.io/docs/migration.html 'It WON’T delete unused columns to protect your data.' we should not have done this right '// Clean up redundant unique indexes'

Option could have been 
1. Check if can't create - stop 
2. Drop what we have figured we want to create?

Please let me know.

**Pull request merged**